Source/WTF/wtf/glib/SysprofAnnotator.h

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright (C) 2024 Igalia, S.L.
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Library General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Library General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Library General Public License
+ *  along with this library; see the file COPYING.LIB.  If not, write to
+ *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301, USA.
+ */
+
+#pragma once
+
+#include <cstdio>
+#include <sysprof-capture.h>
+#include <wtf/HashMap.h>
+#include <wtf/Lock.h>
+#include <wtf/Vector.h>
+#include <wtf/text/ASCIILiteral.h>
+#include <wtf/text/CString.h>
+
+namespace WTF {
+
+class SysprofAnnotator final {
+    WTF_MAKE_NONCOPYABLE(SysprofAnnotator);
+
+    using RawPointerPair = std::pair<const void*, const void*>;
+    using TimestampAndString = std::pair<int64_t, Vector<char>>;
+
+public:
+
+    static SysprofAnnotator& singleton()
+    {
+        static SysprofAnnotator instance;
+        return instance;
+    }
+
+    void instantMark(const char *name, const char *description, ...) WTF_ATTRIBUTE_PRINTF(3, 4)
```

Comment:
```suggestion
    void instantMark(const char* name, const char* description, ...) WTF_ATTRIBUTE_PRINTF(3, 4)
```

In WebKit the asterisks hang on the type, not the variable name. I'm surprised style checker doesn't catch this.

We're moving away from passing C-style strings, so the function parameters should be `std::span<char>`. You can access the underlying data as e.g. `name.data()`.

Reply:
What are the performance implications of using `std::span` here? The Sysprof code is meant to run in very hot paths, the less constructs it uses the better, and I'm not sure `std::span` will help here.

Reply:
`std::span` is basically a pointer plus an integer (its size). You can copy it by value. It's as cheap as anything gets.

For unowned strings like this one, you basically want to use ASCIILiteral whenever possible (because it does bounds checking when accessed), and `std::span` otherwise. And you can't use ASCIILiteral here unless you're willing to touch the Cocoa implementation, so `std::span` is your best bet.

(`CString` is for when you need to own a simple C-style string, but that's not what you need here. And then `String` is WebKit's default full-featured string type, but it's not a C-style string.)

Reply:
I've changed the `name` variable to `std::span`. I couldn't change the others without breaking the `WTF_ATTRIBUTE_PRINTF` attribute.

I think the end result is a straight downgrade from the previous code, both in terms of legibility and in terms of performance, and I don't think security was increased here by any relevant metric. I'll leave it up to other maintainers to decide if they consider this an acceptable code that they wish to maintain.

Reply:
The security benefit is when standard library assertions are enabled (as they are in all Linux distros); it adds bounds checking to ensure WebKit will crash if you try to read an invalid index. I doubt there are any security bugs in your code here, but we're doing it for all of WebKit as a general hardening measure.

Anyway, I agree that keeping `WTF_ATTRIBUTE_PRINTF` is more important than getting rid of `const char *`, so this seems fine to me.